libs/beacon/src/beacon_manager.cpp

```diff
@@ -17,167 +17,617 @@
 //------------------------------------------------------------------------------
 
 #include "beacon/beacon_manager.hpp"
+#include "crypto/ecdsa.hpp"
 
 namespace fetch {
 namespace dkg {
 
-void BeaconManager::Reset(uint64_t cabinet_size, uint32_t threshold)
+bn::G2 BeaconManager::zeroG2_;
+bn::Fr BeaconManager::zeroFr_;
+bn::G2 BeaconManager::group_g_;
+bn::G2 BeaconManager::group_h_;
+
+constexpr char const *LOGGING_NAME = "DKG";
+
+BeaconManager::BeaconManager(CertificatePtr certificate)
+  : certificate_{std::move(certificate)}
 {
-  cabinet_size_ = cabinet_size;
-  threshold_    = threshold;
+  if (certificate_ == nullptr)
+  {
+    auto ptr = new crypto::ECDSASigner();
+    certificate_.reset(ptr);
+    ptr->GenerateKeys();
+  }
 
-  received_shares_.clear();
-  participants_.clear();
+  static std::once_flag flag;
+
+  std::call_once(flag, []() {
+    bn::initPairing();
+    zeroG2_.clear();
+    zeroFr_.clear();
+    group_g_.clear();
+    group_h_.clear();
+
+    // Values taken from TMCG main.cpp
+    const bn::Fp2 g(
+        "1380305877306098957770911920312855400078250832364663138573638818396353623780",
+        "14633108267626422569982187812838828838622813723380760182609272619611213638781");
+    const bn::Fp2 h(
+        "6798148801244076840612542066317482178930767218436703568023723199603978874964",
+        "12726557692714943631796519264243881146330337674186001442981874079441363994424");
+    bn::mapToG2(group_g_, g);
+    bn::mapToG2(group_h_, h);
+  });
+}
 
-  verification_vectors_.resize(cabinet_size);
-  received_shares_.resize(cabinet_size);
+void BeaconManager::GenerateCoefficients()
+{
+  std::vector<bn::Fr> a_i(polynomial_degree_ + 1, zeroFr_), b_i(polynomial_degree_ + 1, zeroFr_);
```

Comment:
style: we don't usually declare multiple on the same line

Reply:
Sure.